### ISSUE TITLE
Tablet hovering impossible without turning on "Tablet" option in input settings.

### ISSUE THREAD

**Issue by Dirtinator** [ppy/osu#30641]
### Type

Game behaviour

### Bug description

After the most recent update (Lazer 2024.1115.1) the game is incapable of properly detecting my tablet unless I drag it. 

If I hover it behaves incorrectly, both as if it had absolute positioning (like a mouse) and partially like normal, jumping between the 2 rapidly.

This has not happened in any previous cases of me playing on Lazer in the past (approximately) year.

The tablet (CTL-472 Wacom) works without issue outside of osu! and this seems to be exclusively when focus is set on the osu! executable.

### Screenshots or videos

https://github.com/user-attachments/assets/49c3745f-f89b-4e17-bac1-85b0fa209115

During the test I was doing square and circular motion to demonstrate the issue. 

### Version

Lazer 2024.1115.1

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/17774108/compressed-logs.zip)

**Comment by 100925afg** [comment]
I also have the same problem.After this update,my tablet(GAOMON 1061PRO) can't move correctly and sometimes can't move to the edge. But it didn't happen at the previous version.
: (

**Comment by 100925afg** [comment]
I also tried to turn off the "high precision mouse",then it fixed(Even though my tablet didn't have supportive driver for osu).
Thanks  for your suggestion!

**Comment by SupDos** [comment]
Possibly related/duplicate of https://github.com/ppy/osu/issues/28223

Try turning off high precision and see if that fixes it

**Comment by SupDos** [comment]
I'm assuming you turned it off in lazer, right? Someone in the dev discord thought they had to change the Windows mouse setting, but changing the setting in lazer fixed it for them

**Comment by Dirtinator** [comment]
Yes the setting was changed in the in-game setting "high precision mouse" and doing so made the game not recognize hovering at all. with it turned on it is in a broken state. Dragging seems to be unaffected.

**Comment by Dirtinator** [comment]
Yeah there also seems to be a separate issue when I have tablet mode enabled in general and hovering though not nearly as major.
The cursor seems to constantly shake very slightly when hovering (tablet mode enabled) and this scales the higher up the pen is held (more distance between the tablet = more violent shaking). the shaking in proportion to the game is miniscule however, less than half a millimeter on the monitor (most likely singular pixels) but when held further it does deviate in position up around 4-5mm.

**Comment by bdach** [comment]
So I guess we turn off SDL3 once more? cc @ppy/team-client

**Comment by Susko3** [comment]
@Dirtinator @100925afg Can you record a video of it being broken again, but have Global Statistics open (<kbd>Ctrl</kbd>-<kbd>F2</kbd>) and showing the `Input` sections.

**Comment by Susko3** [comment]
There are also a few SDL error logs, not sure what those are about:

```
2024-11-15 10:08:59 [verbose]: SDL error log [debug]: Invalid window
2024-11-15 10:09:06 [verbose]: Received beatmap updates 1 updates with last id 8687711
2024-11-15 10:09:14 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:14 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:14 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:14 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:14 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:20 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:20 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:20 [verbose]: SDL error log [debug]: No HID devices found in the system.
2024-11-15 10:09:39 [verbose]: SDL error log [debug]: Invalid window
2024-11-15 10:09:43 [verbose]: SDL error log [debug]: Invalid window
```

**Comment by Dirtinator** [comment]
This is in regards to the hovering not working properly.

https://github.com/user-attachments/assets/c05e5431-da9d-4cbe-b3dc-476cfa14883d

This is in regards to the shaking while hovering (tablet mode)

https://github.com/user-attachments/assets/36f9ca4d-c75e-4995-9fc9-1b4b8e3fd82e

**Comment by Susko3** [comment]
@Dirtinator does hovering work in windowed mode? Try relative mouse mode both enabled and disabled. It should work better when disabled.

**Comment by 100925afg** [comment]
@Susko3 I reinstall my tablet driver ,then it fixed completely.
But I still don't know why it can fixed that bug
(also close the "Tablet","Touch","High precision mouse" selection)

**Comment by 100925afg** [comment]
Now the hovering part also can work(The Driver I use isn't OpenTabletDriver)

**Comment by 100925afg** [comment]
[compressed-logs.zip](https://github.com/user-attachments/files/17774994/compressed-logs.zip)
I don't know if it is useful now...but i still upload my logs
@Susko3

**Comment by Dirtinator** [comment]
@Susko3 Okay that does seem to have changed something. In windowed mode with high precision ON the issue persists however, with high precision OFF the issue disappears completely (and the shaking that was happening with tablet mode is not present). In borderless however the same thing happened as in the original post.

Unfortunately I get worse performance (massive lagspikes) when in windowed/borderless compared to fullscreen but for now the slight shaking isn't causing me any real problems so it's a functional band-aid fix.

**Comment by Susko3** [comment]
To summarise my testing with https://github.com/Teages/vTablet (Windows Ink):

- clicking the pen always works as expected (absolute positioning)
- if relative mode is enabled, hovering will be using relative mode so it wont work as expected. this is by SDL3 design
- if relative mode is disabled, hovering works fine in windowed mode, but doesn't work in fullscreen and borderless
  - i'd say this is an SDL3 bug, or some windows weirdness

The real fix for all this is https://github.com/libsdl-org/SDL/issues/10516.

**Comment by bdach** [comment]
Which means either one of us is going to need to take 2 weeks out to implement this for sdl because they likely don't have enough users to care. Which means, that short term, *we probably revert sdl3 again*.

**Comment by 100925afg** [comment]
> To summarise my testing with https://github.com/Teages/vTablet (Windows Ink):总结一下我在https://github.com/Teages/vTablet （Windows Ink）上的测试：
> 
> * clicking the pen always works as expected (absolute positioning)点击笔总是按预期工作（绝对定位）
> * if relative mode is enabled, hovering will be using relative mode so it wont work as expected. this is by SDL3 design如果启用了相对模式，悬停将使用相对模式，因此它不会像预期的那样工作。这是由SDL3设计的
> * if relative mode is disabled, hovering works fine in windowed mode, but doesn't work in fullscreen and borderless
>   如果禁用相对模式，悬停在窗口模式下可以正常工作，但在全屏和无边框模式下不起作用
>   * i'd say this is an SDL3 bug, or some windows weirdness我想说这是SDL3的bug，或者是一些奇怪的窗口
> 
> The real fix for all this is [libsdl-org/SDL#10516](https://github.com/libsdl-org/SDL/issues/10516).真正的解决方法是 [libsdl-org/SDL#10516](https://github.com/libsdl-org/SDL/issues/10516) 。

I turing off the "Windows Ink"selection in my tablet driver,so maybe this problem can be solved by banned the "Windows Ink Input"?
I can't make sure if my conjecture is correct,but this bug really didn't happen in my device now.

**Comment by Halbour** [comment]
I confirm I also have this problem, using the same Tablet as OP.

**Comment by Dirtinator** [comment]
@Susko3 Sorry if this is pedantic but should I also create a separate issue for what I mentioned in https://github.com/ppy/osu/issues/30641#issuecomment-2478577022?

I can't do so at the moment as I do not have access to my computer but if it would help with organizing I can do so in 2-3 hours unless there is no point.

**Comment by Susko3** [comment]
The shaking is relevant for [OpenTabletDriver](http://opentabletdriver.net/). It's reporting the raw, unfiltered data (as it's designed to do) and the shaking/variance in position is just a natural consequence. You can try raising an issue with them, but they'll likely to tell you to use a filter (and I don't think that's possible with using the in-game `Tablet` settings).

**Comment by Dirtinator** [comment]
The issue has been (temporarily) fixed following the hotfix in Lazer 2024.1115.2

The game now properly detects hovering without issue like it did before this issue.

**Comment by peppy** [comment]
@Susko3 This still looks to be blocking SDL3 rollout for windows (which I'd like to get prioritised to get https://github.com/ppy/osu-framework/pull/6235 in too). I see you have this assigned to yourself – is your goal to attempt to fix this on the SDL3 end?

**Comment by Susko3** [comment]
SDL has recently added windows ink / pen support in https://github.com/libsdl-org/SDL/pull/11826, but I'm [not](https://github.com/libsdl-org/SDL/issues/11843) [confident](https://github.com/libsdl-org/SDL/issues/11844) in their implementation (mostly because it's not strictly following Microsoft documentation, but also because of a lack of testing). Things look to be [improving](https://github.com/libsdl-org/SDL/pull/11869).

I'll try to borrow some testing equipment (graphics tablets, touchscreen monitors) and test out SDL changes.

**Comment by Susko3** [comment]
I've tested this with the Wacom intuos 3 PTZ-930, drivers version `WacomTablet_6.3.15-3`. My tablet works with in game tablet drives, so I had them disabled when testing.

I'm unable to reproduce the original issue, even on the reported version of osu!.
The new SDL3 pen / windows ink handling works really well, there are no problems with relative mode. High precision mouse simply doesn't affect pen handling. Hopefully it will solve this issue, but let's keep it open until OP confirms the fix.

**Comment by Dirtinator** [comment]
@Susko3 i can test if it works on my setup later today most likely. Should I go through the same methodology and use the current, most recent release of the lazer client? If I should pay attention to anything in particular please do mention.

**Comment by bdach** [comment]
@Susko3 how does *that* work? So we're going to just merge changes blind again and revert to sdl2 again if it doesn't work again?

It's been *two times* that moving to sdl3 has been blocked by this issue. Are we really about to try the same thing blind again without *any* testing whatsoever?

**Comment by smoogipoo** [comment]
> try the same thing blind again without any testing

Fwiw, this is a "yes" from me, unless you have any better suggestions. Release -> note any/all issues -> revert if required -> wait for issues to be fixed -> try again.

Perhaps we can add an option to make it a bit easier, but I'd like it to be reset whenever we restart the cycle.
